File: scraper/sources/sarawak.js
'use strict';

// Sarawak eTender Notice — JSP public listing, plain HTML table, no JS required
const axios = require('axios');
const cheerio = require('cheerio');
const { parseDate, inferStatus, nowIso } = require('../utils');

const SOURCE_ID = 4;
const SOURCE_NAME = 'Sarawak';
const BASE_URL = 'https://etendernotice.sarawak.gov.my/etender/public/public_tender_list.jsp';

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
  'Accept': 'text/html,application/xhtml+xml,*/*;q=0.9',
  'Accept-Language': 'ms-MY,ms;q=0.9,en-US;q=0.8',
};

const SKIP_HEADERS = new Set(['no.', 'no', 'tajuk tender', 'tajuk', 'tarikh iklan', 'tarikh tutup',
  'tarikh mula', 'status', 'tindakan', 'title', 'closing date', 'open date', 'action', 'ref no', 'no. rujukan']);

async function* scrape() {
  const now = nowIso();
  let totalYielded = 0;

  try {
    for (let page = 1; page <= 50; page++) {
      const url = page === 1 ? BASE_URL : `${BASE_URL}?page=${page}`;
      const { data } = await axios.get(url, { headers: HEADERS, timeout: 30000 });
      const $ = cheerio.load(data);

      const rows = [];
      $('table tr').each((_, tr) => {
        const tds = $(tr).find('td');
        if (!tds.length) return;
        const cells = tds.map((_, td) => $(td).text().trim()).get();
        if (cells.length < 2) return;

        // Skip header rows
        const firstCell = cells[0].toLowerCase().replace(/[.\s]+/g, ' ').trim();
        if (SKIP_HEADERS.has(firstCell)) return;

        // Title is usually the longest non-date cell; try col 1 then col 0
        let title = cells[1] || cells[0];
        if (!title || title.length < 10) return;

        // Detect date-like columns (dd/mm/yyyy or yyyy-mm-dd)
        const dates = cells.filter(c => /\d{1,2}[\/\-]\d{1,2}[\/\-]\d{4}|\d{4}[\/\-]\d{2}[\/\-]\d{2}/.test(c));
        const closeRaw = dates[dates.length - 1] || null;
        const openRaw  = dates.length > 1 ? dates[0] : null;

        const link = $(tds[1] || tds[0]).find('a').attr('href') || $(tds[0]).find('a').attr('href');
        rows.push({ title, closeRaw, openRaw, link });
      });

      if (rows.length === 0) break;
      console.log(`[${SOURCE_NAME}] page ${page}: ${rows.length} rows`);

      for (const r of rows) {
        const open_date = parseDate(r.openRaw);
        const deadline  = parseDate(r.closeRaw);
        yield {
          source_id: SOURCE_ID, ref: null,
          title: r.title, category: null, ministry: null,
          open_date, deadline, status: inferStatus(open_date, deadline),
          url: r.link ? new URL(r.link, BASE_URL).href : BASE_URL,
          scraped_at: now,
        };
        totalYielded++;
      }

      // Stop if no next-page link
      const hasNext = $('a[href*="page="]').filter((_, a) => {
        const href = $(a).attr('href') || '';
        const m = href.match(/page=(\d+)/);
        return m && parseInt(m[1], 10) === page + 1;
      }).length > 0;
      if (!hasNext && page > 1) break;
    }

    console.log(`[${SOURCE_NAME}] done — ${totalYielded} records`);
  } catch (err) {
    console.error(`[${SOURCE_NAME}] fatal: ${err.message}`);
  }
}

module.exports = { SOURCE_ID, SOURCE_NAME, scrape };

File: scraper/sources/sarawak.py
"""
Sarawak eTender public tender list.
URL: https://etendernotice.sarawak.gov.my/etender/public/public_tender_list.jsp

Plain HTML table — requests + BeautifulSoup.
Skips header rows (th elements) and nav/search text.
"""
import logging
from typing import Iterator

from scraper.utils import make_session, parse_date, infer_status, now_iso

SOURCE_ID = 4
SOURCE_NAME = "Sarawak eTender"
BASE_URL = "https://etendernotice.sarawak.gov.my/etender/public/public_tender_list.jsp"
HOST = "https://etendernotice.sarawak.gov.my"

logger = logging.getLogger(__name__)

# Known header/nav cell text to skip
_SKIP_CELLS = {
    "no", "title", "tajuk", "agency", "agensi", "posted date", "closing date",
    "tarikh tutup", "tarikh iklan", "status", "action", "tindakan",
    "* search for title or agency",
}


def scrape() -> Iterator[dict]:
    session = make_session()
    try:
        import urllib3
        urllib3.disable_warnings()
        resp = session.get(BASE_URL, verify=False)
        resp.raise_for_status()
    except Exception as exc:
        logger.error("%s fetch failed: %s", SOURCE_NAME, exc)
        return

    from bs4 import BeautifulSoup
    soup = BeautifulSoup(resp.text, "lxml")

    for row in soup.select("table tr"):
        # Skip rows that only have <th> cells (header rows)
        if row.find("th") and not row.find("td"):
            continue
        cells = [td.get_text(" ", strip=True) for td in row.find_all("td")]
        if len(cells) < 3:
            continue
        # Skip if first two cells look like headers
        if cells[0].lower().strip() in _SKIP_CELLS or cells[1].lower().strip() in _SKIP_CELLS:
            continue

        link = row.find("a")
        url = HOST + link["href"] if link and link.get("href", "").startswith("/") else BASE_URL

        ref      = cells[0].strip()
        title    = cells[1].strip()
        ministry = cells[2].strip() if len(cells) > 2 else None
        deadline = parse_date(cells[3]) if len(cells) > 3 else None
        open_d   = parse_date(cells[4]) if len(cells) > 4 else None
        now = now_iso()
        yield {
            "source_id": SOURCE_ID,
            "ref":       ref or None,
            "title":     title,
            "ministry":  ministry,
            "deadline":  deadline,
            "open_date": open_d,
            "status":    infer_status(open_d, deadline),
            "url":       url,
            "scraped_at": now,
        }
